lib/ebpd_utils.c

```diff
@@ -0,0 +1,57 @@
+#include <linux/err.h>
+#include <string.h>
+#include "bpf/libbpf.h"
+#include "lib/ebpd_utils.h"
+
+int
+ebpd_load_xdp_prog (const char *filepath, int ifindex, void **handle)
+{
+    struct bpf_object *obj = NULL;
+    struct bpf_prog_load_attr load_attr = {
+        .ifindex	= ifindex,
+        .file           = filepath,
+        .prog_type	= BPF_PROG_TYPE_XDP,
+    };
+    int ret, prog_fd = -1;
+
+    ret = bpf_prog_load_xattr(&load_attr, &obj, &prog_fd);
```

Comment:
In C99 and later you should not forward declare variables unless there is a compelling reason to do so (I know the kernel style predates C99 and sometimes contradicts this). Stated differently, make the scope of a variable as small as possible.

const int ret = bpf_prog_load_xattr(...)

Reply:
I am more of a C programmer than C++! In most cases, this is harmless. Will try to change though (As much as possible!)

Reply:
Unfortunately I won't be around for a F2F meeting this week, but I'm happy to video conference any time.

Understood about C vs C++. Hopefully we can standardize on a subset of C++ which is not overly complicated and actually makes your life easier. In particular try to understand std::unique_ptr (and less commonly std::shared_ptr) since it provides a very flexible way to cleanup objects on shutdown or after an error. I promise that most memory leaks will trivially disappear once you get used to it.

I'm not just complaining to be pedantic. There is lots of value to be had from a consistent style, even if the choice of style can be somewhat arbitrary. The Google C++ style has some nice tooling and a well written style guide (https://google.github.io/styleguide/cppguide.html). You should look into clang-format if you have not already, since it makes some parts of code formatting really easy!

There will still be tons of pure C99 to write for the eBPF programs themselves, so don't despair!